### PR TITLE
[SILGen] Fix a bug where the wrong convention was being used for computing the type of a closure thunk

### PR DESCRIPTION
**Pull request by ahatanaka**
The ObjC selector family convention was being used instead of the C function type convention.

rdar://127090209

### CODE REVIEW

**Comment by ahatanaka** [comment]
@swift-ci please test

**Comment by ahatanaka** [comment]
@swift-ci please test

File: lib/SIL/IR/SILFunctionType.cpp [hunk 1]
```diff
@@ -4109,6 +4109,19 @@ static CanSILFunctionType getUncachedSILFunctionTypeForConstant(
     }
   }
 
+  // The type of the native-to-foreign thunk for a swift closure.
+  if (shouldStoreClangType(TC.getDeclRefRepresentation(constant)) &&
+      constant.isForeign && constant.hasClosureExpr()) {
```
Comment (rjmccall) [inline lib/SIL/IR/SILFunctionType.cpp:4114]: Could you check the foreign + closure conditions first?  They seem like they'll weed things out a lot more than the shouldStoreClangType condition, which I guess is just future-proofing against other kinds of foreign reference.

**Review by rjmccall — APPROVED** [review]
LGTM

**Comment by ahatanaka** [comment]
@swift-ci please test

**Pull request merged**